wpinet/src/main/native/cpp/WebSocketSerializer.h

```diff
@@ -0,0 +1,270 @@
+
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include <wpi/SmallVector.h>
+#include <wpi/SpanExtras.h>
+
+#include "wpinet/WebSocket.h"
+#include "wpinet/uv/Buffer.h"
+
+namespace wpi::detail {
+
+class SerializedFrames {
+ public:
+  SerializedFrames() = default;
+  SerializedFrames(const SerializedFrames&) = delete;
+  SerializedFrames& operator=(const SerializedFrames&) = delete;
+  ~SerializedFrames() { ReleaseBufs(); }
+
+  size_t AddFrame(const WebSocket::Frame& frame, bool server) {
+    if (server) {
+      return AddServerFrame(frame);
+    } else {
+      return AddClientFrame(frame);
+    }
+  }
+
+  size_t AddClientFrame(const WebSocket::Frame& frame);
+  size_t AddServerFrame(const WebSocket::Frame& frame);
+
+  void ReleaseBufs() {
+    for (auto&& buf : m_allocBufs) {
+      buf.Deallocate();
+    }
+    m_allocBufs.clear();
+  }
+
+  SmallVector<uv::Buffer, 4> m_allocBufs;
+  SmallVector<uv::Buffer, 4> m_bufs;
+  size_t m_allocBufPos = 0;
+};
+
+class WebSocketWriteReqBase {
+ public:
+  template <typename Stream, typename Req>
+  int Continue(Stream& stream, std::shared_ptr<Req> req);
+
+  SmallVector<uv::Buffer, 4> m_userBufs;
+  SerializedFrames m_frames;
+  SmallVector<int, 0> m_continueFrameOffs;
+  size_t m_continueBufPos = 0;
+  size_t m_continueFramePos = 0;
+};
+
+template <typename Stream, typename Req>
+int WebSocketWriteReqBase::Continue(Stream& stream, std::shared_ptr<Req> req) {
+  if (m_continueBufPos >= m_frames.m_bufs.size()) {
+    return 0;  // nothing more to send
+  }
+
+  // try writing everything remaining
+  std::span bufs = std::span{m_frames.m_bufs}.subspan(m_continueBufPos);
+  int sentBytes = stream.TryWrite(bufs);
+  if (sentBytes < 0) {
+    return sentBytes;  // error
+  }
+
+  int numBytes = 0;
+  for (auto&& buf : bufs) {
+    numBytes += buf.len;
+  }
+  if (sentBytes == numBytes) {
+    m_continueBufPos = m_frames.m_bufs.size();
+    return 0;  // nothing more to send
+  }
+
+  // we didn't send everything; deal with the leftovers
+
+  // figure out what the last (partially) frame sent actually was
+  auto offIt = m_continueFrameOffs.begin() + m_continueFramePos;
+  auto offEnd = m_continueFrameOffs.end();
+  while (offIt != offEnd && *offIt < sentBytes) {
+    ++offIt;
+  }
+  assert(offIt != offEnd);
+
+  // build a list of buffers to send as a normal write:
+  SmallVector<uv::Buffer, 4> writeBufs;
+  auto bufIt = bufs.begin();
+  auto bufEnd = bufs.end();
+
+  // start with the remaining portion of the last buffer actually sent
+  int pos = 0;
+  while (bufIt != bufEnd && pos < sentBytes) {
+    pos += (bufIt++)->len;
+  }
+  if (bufIt != bufs.begin() && pos != sentBytes) {
+    writeBufs.emplace_back(
+        wpi::take_back((bufIt - 1)->bytes(), pos - sentBytes));
+  }
+
+  // continue through the last buffer of the last partial frame
+  while (bufIt != bufEnd && pos < *offIt) {
+    pos += bufIt->len;
+    writeBufs.emplace_back(*bufIt++);
+  }
+  ++offIt;
```

Comment:
Could this potentially move `offIt` to be at `offEnd`? (This causes problems for line 117 (checking `pos` against `*offIt`) and line 121 (advancing `offIt` again, potentially to `offEnd + 1`))